src/ApplicationInsights.Kubernetes/K8sHttpClient/KubeHttpClientSettingsBase.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            string tokenFilePath = GetTokenFilePath();
+            if (!File.Exists(tokenFilePath))
+            {
+                throw new FileNotFoundException("Token file doesn't exist for kubernetes query.", tokenFilePath);
+            }
+            return File.ReadAllText(tokenFilePath);
+        }
+
+        protected abstract string GetTokenFilePath();
+        protected abstract string GetCertFilePath();
+
+        protected static string FetchQueryNamespace(string pathToNamespace)
+        {
+            if (!File.Exists(pathToNamespace))
+            {
+                throw new FileNotFoundException("File contains namespace does not exist.", pathToNamespace);
+            }
+            return File.ReadAllText(pathToNamespace);
+        }
+
+        /// <summary>
+        /// Verify server certificate within returned HttpResponse to prevent MITM attack.
+        /// </summary>
+        /// <param name="httpRequestMessage">The httpRequestMessage returned.</param>
+        /// <param name="serverCertVerifier">The server certificate.</param>
+        /// <param name="chain">The X509Chain.</param>
+        /// <param name="policyErrors">SslPolicyErrors.</param>
+        /// <param name="clientCertVerifier">Client certificate.</param>
+        /// <returns>Returns true when server certificate is valid.</returns>
+        internal bool VerifyServerCertificate(HttpRequestMessage httpRequestMessage, ICertificateVerifier serverCertVerifier, X509Chain chain, SslPolicyErrors policyErrors, ICertificateVerifier clientCertVerifier)
+        {
+            Arguments.IsNotNull(clientCertVerifier, nameof(clientCertVerifier));
+            X509Certificate2 serverCert = serverCertVerifier.Certificate;
+            _logger.LogDebug("Server certification custom validation callback.");
+            _logger.LogTrace(httpRequestMessage?.ToString());
+            _logger.LogTrace(chain?.ToString());
+            _logger.LogTrace(policyErrors.ToString());
+            _logger.LogTrace("ServerCert:" + Environment.NewLine + serverCert);
+
+            try
+            {
+                // Verify Issuer. Issuer field is case-insensitive.
+                if (!string.Equals(clientCertVerifier.Issuer, serverCertVerifier.Issuer, StringComparison.OrdinalIgnoreCase))
+                {
+                    _logger.LogError(Invariant($"Issuer are different for server certificate and the client certificate. Server Certificate Issuer: {clientCertVerifier.Issuer}, Client Certificate Issuer: {serverCertVerifier.Issuer}"));
+                    return false;
+                }
+                else
+                {
+                    _logger.LogDebug(Invariant($"Issuer validation passed: {serverCertVerifier.Issuer}"));
+                }
+
+                // Server certificate is not expired.
+                DateTime now = DateTime.Now;
+                if (serverCertVerifier.NotBefore > now || serverCertVerifier.NotAfter.AddDays(1) <= now)
```

Comment:
Why 1 day of padding?

Reply:
Good catch. I can't remember the exact reason now. But here is my guess -
I thought the NotBefore / NotAfter is always aligned to 12:00am, which makes the valid case be:
NotBefore<=now<NotAfter+24hrs, then reverse the expression to see the invalid case, which is the code.
Well, problem is, NotBefore / NotAfter is not aligned to 12:00am, the valid case should be:
NotBefore <= now <= NotAfter. Reverse it should be:
NotBefore > now || NotAfter < now.

I'll fix it.